src/main/java/teammates/common/util/Url.java

```diff
@@ -2,14 +2,12 @@
 
 public class Url {
 
+    protected final String baseUrl;
     protected String urlString;
 
-    public Url(String url) {
-        this.urlString = url.replace(getAppUrl(), ""); // force the URL to be relative
-    }
-
-    protected String getAppUrl() {
-        return Config.APP_URL;
+    public Url(String baseUrl, String url) {
```

Comment:
Turns out I can't really make it defensive without adding unnecessary restriction, because it will still work if the base URL has trailing slash but the path doesn't have leading slash. As for the leading `http://` or `https://`, it also adds an unnecessary restriction because this class can actually be used for `file:///` links (I just thought of it).
The parameters are that way for a purpose. If the 2nd param starts with baseUrl, it will be trimmed to form the path only. Otherwise it will still work. Examples:
```java
new Url("http://www.google.com", "/page").toString() = "/page";
new Url("http://www.google.com", "/page").toAbsoluteString() = "http://www.google.com/page";
new Url("http://www.google.com", "http://www.google.com/page").toString() = "/page";
new Url("http://www.google.com", "http://www.google.com/page").toAbsoluteString() = "http://www.google.com/page";
```
There are use cases for all of them.